src/cryptonote_basic/miner.cpp

```diff
@@ -846,40 +846,81 @@ namespace cryptonote
+            if (boost::filesystem::is_regular_file(power_supply_type_path))
+            {
+              std::string power_supply_type;
+              epee::file_io_utils::load_file_to_string(power_supply_type_path, power_supply_type);
+
+              // If there is an AC adapter that's present and online we can break early
+              if (boost::starts_with(power_supply_type, "Mains"))
+              {
+                std::string power_supply_online_path = power_supply_path.string() + "/online";
+                if (boost::filesystem::is_regular_file(power_supply_online_path))
+                {
+                  std::string power_supply_online;
+                  epee::file_io_utils::load_file_to_string(power_supply_online_path, power_supply_online);
+
+                  if (boost::starts_with(power_supply_online, "1"))
+                  {
+                    return boost::logic::tribool(false);
+                  }
+                  else
+                  {
+                    continue;
+                  }
+                }
+              }
+              else if (boost::starts_with(power_supply_type, "Battery"))
+              {
+                std::string power_supply_status_path = power_supply_path.string() + "/status";
+                if (boost::filesystem::is_regular_file(power_supply_status_path))
+                {
+                  std::string power_supply_status;
+                  epee::file_io_utils::load_file_to_string(power_supply_status_path, power_supply_status);
+
+                  // Possible status are Charging, Full, Discharging, Not Charging, and Unknown
+                  // We are only need to handle negative states right now
+
+                  if (is_plugged == boost::logic::indeterminate &&
+                      (boost::starts_with(power_supply_status, "Charging") || boost::starts_with(power_supply_status, "Full")))
+                  {
+                    is_plugged = boost::logic::tribool(true);
+                  }
+                  else
+                  {
+                    is_plugged = boost::logic::tribool(false);
+                  }
+                }
+              }
+            }
+          }
         }
       }
 
-      if( power_supply_path.empty() )
-      {
-        LOG_ERROR("Couldn't find battery/power status file, can't determine if plugged in!");
-        return boost::logic::tribool(boost::logic::indeterminate);;
-      }
-
-      std::ifstream power_stream(power_supply_path);
-      if( power_stream.fail() )
-      {
-        LOG_ERROR("failed to open '" << power_supply_path << "'");
-        return boost::logic::tribool(boost::logic::indeterminate);;
+      if (is_plugged == boost::logic::indeterminate) {
+        LOG_ERROR("couldn't query power status from " << power_supply_class_path);
       }
-
-      return boost::logic::tribool( (power_stream.get() != '1') );
+      return is_plugged;
```

Comment:
So if it's plugged in, why you are returning true for a function that is being queried as to whether being run on battery?